gh-140138: NOGIL: add atomic owner refcount to PyInterpreterState; finalize on last tstate

ashm-dev opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNNN: Summary of the changes made
```

Where: gh-NNNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNNNN)
```

Where: [X.Y] is the branch name, for example: [3.13].

GH-NNNNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-140138 -->
* Issue: gh-140138
<!-- /gh-issue-number -->

ashm-dev committed:
fix tsan

ashm-dev committed:
add blurb

ZeroIntensity commented on Include/internal/pycore_interp_structs.h:
```diff
@@ -793,6 +787,7 @@ struct _is {
     int finalizing;
 
     uintptr_t last_restart_version;
+    Py_ssize_t _owners;
```
I don't think we want the `_` prefix; this is already a private API. Also, this should be added to the end of the struct, not the beginning.

ashm-dev replied:
fixed

ZeroIntensity commented on Misc/NEWS.d/next/Core_and_Builtins/2025-10-16-20-06-33.gh-issue-140138.a_fhrh.rst:
This is really long for a news entry. I think having only the first sentence is fine, since users don't understand or care about technical details when reading the changelog.

ashm-dev replied:
fixed

ZeroIntensity commented on Python/pystate.c:
```diff
@@ -481,7 +484,24 @@ free_interpreter(PyInterpreterState *interp)
+}
+
+static void
+_finalize_and_free_interpreter(PyInterpreterState *interp)
```
The convention in this file is to omit a `_` prefix for static functions. Also, this isn't really "finalizing" the interpreter -- finalization is really expensive (think `Py_EndInterpreter` or `Py_Finalize`), while freeing the interpreter is not. Let's call this something else.

ashm-dev replied:
fixed

ZeroIntensity commented on Python/pystate.c:
```diff
@@ -481,7 +484,24 @@ free_interpreter(PyInterpreterState *interp)
+_finalize_and_free_interpreter(PyInterpreterState *interp)
+{
+    _Py_qsbr_fini(interp);
+    _PyObject_FiniState(interp);
```
It's fine to keep these in `PyInterpreterState_Delete`. We only care about the interpreter structure being valid.

ashm-dev replied:
fixed

ZeroIntensity commented on Python/pystate.c:
```diff
@@ -1412,6 +1431,7 @@ static void
 free_threadstate(_PyThreadStateImpl *tstate)
 {
     PyInterpreterState *interp = tstate->base.interp;
+
```
Stray newline change.

ashm-dev replied:
fixed

ZeroIntensity commented on Python/pystate.c:
```diff
@@ -955,11 +976,9 @@ PyInterpreterState_Delete(PyInterpreterState *interp)
-    _Py_qsbr_fini(interp);
-
-    _PyObject_FiniState(interp);
+    interp->finalizing = 1;
```
This is a misuse of the `finalizing` field. `finalizing` should be set when we're cleaning up Python-level state, which has already happened.

ashm-dev replied:
deleted

ashm-dev committed:
Reorder and add `_owners` field

Move the `_owners` field to be after `finalizing` and add it back to the
struct. It was removed in a previous commit and is needed.

ashm-dev committed:
Fix typo in interpreter owner atomic variable

Fixes a typo where the atomic variable `_owners` was incorrectly named
and should have been `owners`.

ashm-dev committed:
Fix news

ashm-dev committed:
Rename interpreter cleanup functions

Rename functions to be more descriptive of their actions.

ashm-dev committed:
Refactor interpreter cleanup

Move the cleanup logic into PyInterpreterState_Delete and remove the
now-unused helper function.

ashm-dev committed:
Rename _owners to owners in interp struct

The field name was inconsistent with other usages.

ashm-dev committed:
fix

ashm-dev committed:
Stop setting `interp->finalizing`

ZeroIntensity reviewed (commented):
Thanks, this PR looks good if we decide to fix this, but I'm worried about whether this is even a bug in the first place. See [my comment](https://github.com/python/cpython/issues/140138#issuecomment-3413241506) on the issue. Please let me know if I missed something in my analysis.

ashm-dev committed:
Merge remote-tracking branch 'upstream/main' into gh-140138
